contracts/src/v0.8/dev/automation/2_1/KeeperRegistryLogicA2_1.sol

```diff
@@ -104,33 +104,66 @@ contract KeeperRegistryLogicA2_1 is
       return (false, bytes(""), UpkeepFailureReason.INSUFFICIENT_BALANCE, 0, upkeep.executeGas, 0, 0);
     }
 
-    upkeepNeeded = true;
-
     bytes memory callData;
     if (triggerType == Trigger.CONDITION) {
       callData = abi.encodeWithSelector(CHECK_SELECTOR, checkData);
     } else {
       callData = abi.encodeWithSelector(CHECK_LOG_SELECTOR, checkData);
     }
     gasUsed = gasleft();
-    (upkeepNeeded, performData) = upkeep.target.call{gas: s_storage.checkGasLimit}(callData);
+    (bool success, bytes memory result) = upkeep.target.call{gas: s_storage.checkGasLimit}(callData);
     gasUsed = gasUsed - gasleft();
-    if (!upkeepNeeded) {
-      upkeepFailureReason = UpkeepFailureReason.TARGET_CHECK_REVERTED;
-    } else {
-      (upkeepNeeded, performData) = abi.decode(performData, (bool, bytes));
-      if (!upkeepNeeded)
+
+    if (!success) {
+      // User's target check reverted. We capture the revert data here and pass it within performData
+      if (result.length > s_storage.maxRevertDataSize) {
         return (
           false,
           bytes(""),
-          UpkeepFailureReason.UPKEEP_NOT_NEEDED,
+          UpkeepFailureReason.REVERT_DATA_EXCEEDS_LIMIT,
           gasUsed,
           upkeep.executeGas,
           fastGasWei,
           linkNative
         );
+      }
+      return (
+        upkeepNeeded,
+        result,
+        UpkeepFailureReason.TARGET_CHECK_REVERTED,
+        gasUsed,
+        upkeep.executeGas,
+        fastGasWei,
+        linkNative
+      );
+    }
+
+    try abi.decode(result, (bool, bytes)) returns (bool un, bytes pd) {
+      upkeepNeeded = un;
+      performData = pd;
+    } catch {
```

Comment:
wait is this actually a thing? I thought you could only try/catch external function calls.

Reply:
ah i get a compile error
`TypeError: Try can only be used with external function calls and contract creation calls.`